pkg/bloombuild/planner/queue/queue.go

```diff
@@ -0,0 +1,171 @@
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/dskit/services"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/grafana/loki/v3/pkg/queue"
+	"github.com/grafana/loki/v3/pkg/util"
+)
+
+type Task interface {
+	Tenant() string
+	Table() string
+	ID() string
+}
+
+// Queue is a wrapper of queue.RequestQueue that uses the file system to store the pending tasks.
+// When a task is enqueued, it's stored in the file system and recorded ad pending.
+// When it's dequeued, it's removed from the queue but kept in FS until removed.
+type Queue struct {
+	services.Service
+
+	queue *queue.RequestQueue
+	// pendingTasks is a map of task ID to the file where the task is stored.
+	pendingTasks sync.Map
+	activeUsers  *util.ActiveUsersCleanupService
+
+	cfg    Config
+	logger log.Logger
+
+	// Subservices manager.
+	subservices        *services.Manager
+	subservicesWatcher *services.FailureWatcher
+}
+
+func NewQueue(logger log.Logger, cfg Config, limits Limits, metrics *Metrics) (*Queue, error) {
+	tasksQueue := queue.NewRequestQueue(cfg.MaxQueuedTasksPerTenant, 0, limits, metrics)
+
+	// Clean metrics for inactive users: do not have added tasks to the queue in the last 1 hour
+	activeUsers := util.NewActiveUsersCleanupService(5*time.Minute, 1*time.Hour, func(user string) {
+		metrics.Cleanup(user)
+	})
+
+	svcs := []services.Service{tasksQueue, activeUsers}
+	subservices, err := services.NewManager(svcs...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create subservices manager: %w", err)
+	}
+	subservicesWatcher := services.NewFailureWatcher()
+	subservicesWatcher.WatchManager(subservices)
+
+	q := &Queue{
+		queue:              tasksQueue,
+		activeUsers:        activeUsers,
+		cfg:                cfg,
+		logger:             logger,
+		subservices:        subservices,
+		subservicesWatcher: subservicesWatcher,
+	}
+
+	q.Service = services.NewIdleService(q.starting, q.stopping)
+
+	return q, nil
+}
+
+func (q *Queue) starting(ctx context.Context) error {
+	if err := services.StartManagerAndAwaitHealthy(ctx, q.subservices); err != nil {
+		return fmt.Errorf("failed to start task queue subservices: %w", err)
+	}
+
+	return nil
+}
+
+func (q *Queue) stopping(_ error) error {
+	if err := services.StopManagerAndAwaitStopped(context.Background(), q.subservices); err != nil {
+		return fmt.Errorf("failed to stop task queue subservices: %w", err)
+	}
+
+	return nil
+}
+
+func (q *Queue) GetConnectedConsumersMetric() float64 {
+	return q.queue.GetConnectedConsumersMetric()
+}
+
+func (q *Queue) NotifyConsumerShutdown(consumer string) {
+	q.queue.NotifyConsumerShutdown(consumer)
+}
+
+func (q *Queue) RegisterConsumerConnection(consumer string) {
+	q.queue.RegisterConsumerConnection(consumer)
+}
+func (q *Queue) UnregisterConsumerConnection(consumer string) {
+	q.queue.UnregisterConsumerConnection(consumer)
+}
+
+// Enqueue adds a task to the queue.
+func (q *Queue) Enqueue(tenant string, task Task, successFn func()) error {
```

Comment:
Do we still need the interface when the Queue is not generic? Can we use `*QueueTask` instead, and move the QueueTask into this package?

```suggestion
func (q *Queue) Enqueue(tenant string, task *QueueTask, successFn func()) error {
```

Reply:
Not a big deal but:
- We set the `queueTime` only once when we add it to the queue for the first time.
- The queueTask has a `resultsChannel` field that has nothing to do with the queue, but with the rest of the queue impl.

I think  we should leave the queue where it's now. But no strong opinions.

Reply:
Ok with it.